src/farm/library.cairo

```diff
@@ -50,15 +78,274 @@ namespace CarbonableFarmer {
-        ) -> (reward_token_address: felt) {
-        let (reward_token_address) = reward_token_address_.read();
-        return (reward_token_address,);
+    func carbonable_minter_address{syscall_ptr: felt*, pedersen_ptr: HashBuiltin*, range_check_ptr}(
+        ) -> (carbonable_minter_address: felt) {
+        let (carbonable_project_address) = carbonable_project_address_.read();
+        let (carbonable_minter_address) = ICarbonableProject.owner(carbonable_project_address);
+        return (carbonable_minter_address=carbonable_minter_address,);
+    }
+
+    func is_locked{syscall_ptr: felt*, pedersen_ptr: HashBuiltin*, range_check_ptr}() -> (
+        status: felt
+    ) {
+        alloc_locals;
+
+        // [Compute] Absolute locked time
+        let (current_time) = get_block_timestamp();
+        let (start) = start_.read();
+        let (unlocked_duration) = unlocked_duration_.read();
+        let locked_time = start + unlocked_duration;
+
+        // [Compute] Absolute end time
+        let (period_duration) = period_duration_.read();
+        let end = start + period_duration;
+
+        // [Evaluate] Boudaries and current time
+        let is_before_locked = is_le(current_time, locked_time);
+        let is_after_locked = is_le(end, current_time);
+        let over = is_not_zero(is_before_locked + is_after_locked);
+        let status = 1 - over;
+        
+        return (status=status,);
+    }
+
+    func total_locked{syscall_ptr: felt*, pedersen_ptr: HashBuiltin*, range_check_ptr}() -> (
+        balance: Uint256
+    ) {
+        let (contract_address) = get_contract_address();
+        let (carbonable_project_address) = carbonable_project_address_.read();
+
+        let (balance) = IERC721.balanceOf(
+            contract_address=carbonable_project_address, owner=contract_address
+        );
+
+        return (balance=balance,);
+    }
+
+    func share{syscall_ptr: felt*, pedersen_ptr: HashBuiltin*, range_check_ptr}(
+        address: felt, precision: felt
+    ) -> (share: Uint256) {
+        alloc_locals;
+
+        let (total_deposite) = count(address);
+        let (total_contract) = total_locked();
+
+        let dividend = precision * total_deposite;
+        let dividend_uint256 = Uint256(low=dividend, high=0);
+        let (share, _) = uint256_unsigned_div_rem(dividend_uint256, total_contract);
+        return (share=share,);
+    }
+
+    func registred_owner_of{syscall_ptr: felt*, pedersen_ptr: HashBuiltin*, range_check_ptr}(
+        token_id: Uint256
+    ) -> (address: felt) {
+        // [Check] Uint256 compliance
+        with_attr error_message("CarbonableFarmer: token_id is not a valid Uint256") {
+            uint256_check(token_id);
+        }
+        let (address) = registration_.read(token_id);
+        return (address=address,);
+    }
+
+    //
+    // Externals
+    //
+
+    func start_period{syscall_ptr: felt*, pedersen_ptr: HashBuiltin*, range_check_ptr}(
```

Comment:
To start and stop a period, I think we want to control/ensure that it's only the admin which is able to do it. 
Do we ?

To ensure this, maybe add a test in "integration test", where the persona "anyone" try to start and stop a period and failed to do it.

Reply:
You are right, these externals are protected but they were not tested.
I added tests for both yielder and offseter